File: homework3/danny_tat/q3_priority_queue.py
class PriorityQueue:
    def __init__(self):
        self.arr = []  # list of (string, int) pairs

    def top(self):
        # Returns the highest-priority (first) element
        return self.arr[0][0] if self.arr else None

    def insert(self, x, weight):
        # Adds string x with priority weight
        self.arr.append((x, weight))
        self._sift_up(len(self.arr) - 1)

    def remove(self):
        # Removes the highest-priority element
        if not self.arr:
            return
        self.arr[0], self.arr[-1] = self.arr[-1], self.arr[0]
        self.arr.pop()
        if self.arr:
            self._sift_down(0)

    def _sift_up(self, i):
        while i > 0:
            parent = (i - 1) // 2
            if self.arr[i][1] > self.arr[parent][1]:  # max heap on weight
                self.arr[i], self.arr[parent] = self.arr[parent], self.arr[i]
                i = parent
            else:
                break

    def _sift_down(self, i):
        n = len(self.arr)
        while True:
            largest = i
            left, right = 2 * i + 1, 2 * i + 2
            if left < n and self.arr[left][1] > self.arr[largest][1]:
                largest = left
            if right < n and self.arr[right][1] > self.arr[largest][1]:
                largest = right
            if largest == i:
                break
            self.arr[i], self.arr[largest] = self.arr[largest], self.arr[i]
            i = largest


# Time: 20 min

File: homework3/steven_law/q3_PriorityQueue.cpp
// Time spent: 18 mins 47 seconds

#include <iostream>
#include <vector>
#include <string>
using namespace std;

class PriorityQueue {
private:
    vector<pair<string, int>> arr;

    // bubble up at index i
    void bubbleUp(int i){
        while(i > 0){
            int parent = (i - 1) / 2;
            if(arr[i].second > arr[parent].second){ // max heap so flipped from heap impl
                swap(arr[i], arr[parent]);
                i = parent;
            } else {
                break;
            }
        }
    }

    // bubble down from index i. 
    void bubbleDown(int i){
        int n = arr.size();
        while(true){
            int left = 2*i + 1;
            int right = 2*i + 2;
            int largest = i;

            if(left < n && arr[left].second > arr[largest].second) largest = left;
            if(right < n && arr[right].second > arr[largest].second) largest = right;

            if(largest == i) break;
            swap(arr[i], arr[largest]);
            i = largest;
        }
    }

public:
    // returns the highest priority (first) element in the PQ
    int top(){
        if(arr.empty()) return -1;
        return arr[0].second;
    }

    // helper to grab the string since top() returns the weight
    string topElement(){
        if(arr.empty()) return "";
        return arr[0].first;
    }

    // adds string x to the PQ with priority weight
    void insert(string x, int weight){
        arr.push_back({x, weight});
        bubbleUp(arr.size() - 1);
    }

    // removes the highest priority (first) element in the PQ
    void remove(){
        if(arr.empty()) return;
        arr[0] = arr.back();
        arr.pop_back();
        if(!arr.empty()) bubbleDown(0);
    }
};

int main(){
    PriorityQueue pq;
    pq.insert("low", 1);
    pq.insert("high", 10);
    pq.insert("mid", 5);
    pq.insert("highest", 20);

    cout << "Top: " << pq.topElement() << " priority " << pq.top() << " (Expected: highest 20)" << endl;
    pq.remove();
    cout << "After remove: " << pq.topElement() << " priority " << pq.top() << " (Expected: high 10)" << endl;
    pq.remove();
    cout << "After remove: " << pq.topElement() << " priority " << pq.top() << " (Expected: mid 5)" << endl;

    return 0;
}
